client.go

```diff
@@ -189,34 +189,100 @@ func (c *Client) exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err erro
 // If the received message contains a TSIG record the transaction
 // signature is verified.
 func (co *Conn) ReadMsg() (*Msg, error) {
-	var p []byte
+	p, err := co.ReadMsgBytes(nil)
+	if err != nil {
+		return nil, err
+	}
+
 	m := new(Msg)
-	if _, ok := co.Conn.(*net.TCPConn); ok {
-		p = make([]byte, MaxMsgSize)
+	if err := m.Unpack(p); err != nil {
+		return nil, err
+	}
+	if t := m.IsTsig(); t != nil {
+		if _, ok := co.TsigSecret[t.Hdr.Name]; !ok {
+			return m, ErrSecret
+		}
+		// Need to work on the original message p, as that was used to calculate the tsig.
+		err = TsigVerify(p, co.TsigSecret[t.Hdr.Name], co.tsigRequestMAC, false)
+	}
+	return m, err
+}
+
+// ReadMsgBytes reads DNS packet, parses and fills wire header (passing nil
+// would cause skipping that). Returns message in byte format to parse with
+// Msg.Unpack later on.
+//
+// Note that this function would not be able to report TSIG error or
+// check it got actual DNS payload.
+func (co *Conn) ReadMsgBytes(hdr *Header) ([]byte, error) {
```

Comment:
So... What if we would really makes this only read 12 (or 14 in case of tcp) bytes and let it return the Conn, the bytes read and an error? Maybe the buffer should also by given to the function, so it is really without any internal allocations? So if you get an error you just discard and write FormErr (maybe from within the lib already). If thinks look OK we can parse the bytes to a real header and assemble the packet...